src/Microsoft.Management.Deployment/CatalogPackage.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include <mutex>
+#include <AppInstallerRepositorySource.h>
+#include <AppInstallerRepositorySearch.h>
+#include "CatalogPackage.h"
+#include "CatalogPackage.g.cpp"
+#include "PackageCatalog.h"
+#include "PackageVersionInfo.h"
+#include "PackageVersionId.h"
+#include <wil\cppwinrt_wrl.h>
+
+namespace winrt::Microsoft::Management::Deployment::implementation
+{
+    void CatalogPackage::Initialize(
+        std::shared_ptr<const ::AppInstaller::Repository::ISource> source,
+        std::shared_ptr<::AppInstaller::Repository::IPackage> package)
+    {
+        m_source = std::move(source);
+        m_package = std::move(package);
+    }
+    hstring CatalogPackage::Id()
+    {
+        return winrt::to_hstring(m_package->GetProperty(::AppInstaller::Repository::PackageProperty::Id).get());
+    }
+    hstring CatalogPackage::Name()
+    {
+        return winrt::to_hstring(m_package->GetProperty(::AppInstaller::Repository::PackageProperty::Name));
+    }
+    Microsoft::Management::Deployment::PackageVersionInfo CatalogPackage::InstalledVersion()
+    {
+        std::scoped_lock lock(m_installedVersionMutex);
```

Comment:
>std::scoped_lock lock(m_installedVersionMutex); [](http://example.com/codeflow?start=8&length=47)

I don't think that it is really a performance difference, but from a lines of code perspective using `std::call_once` would be better for these.  Only 1 member instead of two in the header, and then less bookkeeping for every use, for instance here it would be:
```
        std::call_once(m_installedVersionOnceFlag,
        [&]()
        {
std::shared_ptr<::AppInstaller::Repository::IPackageVersion> installedVersion = m_package.get()->GetInstalledVersion();
            if (installedVersion)
            {
                auto installedVersionImpl = winrt::make_self<wil::details::module_count_wrapper<
                    winrt::Microsoft::Management::Deployment::implementation::PackageVersionInfo>>();
                installedVersionImpl->Initialize(std::move(installedVersion));
                m_installedVersion = *installedVersionImpl;
            }
        });
        return m_installedVersion;
```

Reply:
Fancy.